packages/loader/container-loader/src/connectionStateHandler.ts

```diff
@@ -59,21 +68,63 @@ export class ConnectionStateHandler extends EventEmitterWithErrorHandling<IConne
         private readonly logger: ITelemetryLogger,
     ) {
         super();
+        this.prevClientLeftTimer = new Timer(
+            // Default is 90 sec for which we are going to wait for its own "leave" message.
+            this.handler.maxClientLeaveWaitTime ?? 90000,
+            () => {
+                this.clientLeaveWaitEnded(false);
+            },
+        );
     }
 
-    public receivedAddMemberEvent(clientId: string, quorum: Quorum) {
+    // This is true when this client submitted any ops.
+    public clientSentOps(connectionMode: ConnectionMode) {
+        this._clientSentOps = true;
+        this.clientConnectionMode = connectionMode;
+    }
+
+    public receivedAddMemberEvent(clientId: string) {
         // This is the only one that requires the pending client ID
         if (clientId === this.pendingClientId) {
+            // Wait for previous client to leave the quorum before firing "connected" event.
+            if (this.prevClientLeftTimer.hasTimer) {
+                this.waitEvent = PerformanceEvent.start(this.logger, {
+                    eventName: "WaitBeforeClientLeave",
+                    waitOnClientId: this._clientId,
+                    hadOutstandingOps: this.handler.shouldClientJoinWrite(),
+                });
+                this.waitingClientId = clientId;
+            } else {
+                this.setConnectionState(ConnectionState.Connected);
+            }
+        }
+    }
+
+    private clientLeaveWaitEnded(leaveReceived: boolean) {
+        // Move to connected state only if there was a client waiting for client leave. It may happen that
+        // during wait, the waiting client again got Disconnected/Connecting and then we don't want to
+        // move to connected state here for a non waiting client.
+        if (this.waitingClientId !== undefined) {
```

Comment:
I think we can remove waitingClientId memeber, and instead use this check here:

if (this.pendingClientId !== this.clientId && quorum.has(this.pendingClientId))

This basically tells us that we are in connecting state (first part of the check) and that we already observed our own join op.

This would require quorum to be available at this layer (a minus), but it already has access to it.
Pro - less state to track, less likely have bugs.

The next potential step would be for this API to add to such IF condition !this.prevClientLeftTimer.hasTimer test. That way it can be called on any chance of state, i.e. receivedAddMemberEvent() can simply call into here unconditionally (and do nothing else!). we need to double check that it's possible (I do not know what to do with this.waitEvent in this picture, and if it needs some consideration), but it moves us toward more state machinery this is encapsulated in one place, and thus easier to review and test.

P.S. this assumes that prevClientLeftTimer.hasTimer is false the moment it runs timer callback. I do not know if that's the case.

Reply:
Yes, we can remove waitingClientId with this check.

Reply:
Fixes. Removed waitingClientId.
"this assumes that prevClientLeftTimer.hasTimer is false the moment it runs timer callback.  - This is true.

Reply:
Also I noticed we had a wrong check in Connecting too even in previous code.
 quorum.has(this.pendingClientId)) is wrong.  quorum..getMember(this.pendingClientId) !== undefined or  quorum.getMembers().has(this.pendingClientId)) is the right check to check if the client is in the quorum.